### PR TITLE
Add BindingOptions.MethodInterceptor

### PR DESCRIPTION
**Pull request by joaompneves**
Support Classes serialization when before only Structs were allowed. Following Structs serialization criteria already implemented (serialize public fields), object classes can now also be returned to the JS side.

Added mechanism to intercept execution of managed methods. This introduces support for use cases such as logging or decoration/inspection of objects returned.

### CODE REVIEW

**Comment by amaitland** [comment]
> Support Classes serialization when before only Structs were allowed. Following Structs serialization criteria already implemented (serialize public fields), object classes can now also be returned to the JS side.

Are there two distinct changes rolled up into this `PR`?

> Added mechanism to intercept execution of managed methods. This introduces support for use cases such as logging or decoration/inspection of objects returned.

Is this really the responsibility of `CefSharp`? Seems like a you'd be better to implement a decorator pattern in your own code.

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs [hunk 1]
```diff
@@ -13,11 +13,23 @@
...
 using CefSharp.Wpf.Example.ViewModels;
 using System.IO;
+using System;
 
 namespace CefSharp.Wpf.Example.Views
 {
     public partial class BrowserTabView : UserControl
     {
+
+        private class CustomInterceptor : IInterceptor
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs]: One class per file please.

File: CefSharp/BindingOptions.cs [hunk 2]
```diff
@@ -30,6 +30,11 @@ public static BindingOptions DefaultBinder
         /// </summary>
         public IBinder Binder { get; set; }
 
+        /// <summary>
+        /// interceptor used for intercepting calls to methods
+        /// </summary>
+        public IInterceptor Interceptor { get; set; }
```
Comment (amaitland) [inline CefSharp/BindingOptions.cs]: Comment lacks detail, it wasn't until I looked closer at the usage example until I realized what it was for. Comment needs to be improved.

File: CefSharp/Internals/JavascriptObjectRepository.cs [hunk 3]
```diff
@@ -187,7 +188,14 @@ public bool TryCallMethod(long objectId, string name, object[] parameters, out o
                         }
                     }
 
-                    result = method.Function(obj.Value, parameters);
+                    if (obj.Interceptor != null)
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptObjectRepository.cs]: This should be inverted, preferable to evaluate `==` first.

```c#
if (obj.Interceptor == null) 
{
    result = method.Function(obj.Value, parameters);
}
else 
{
    result = obj.Interceptor.Intercept(() => method.Function(obj.Value, parameters), method.ManagedName);
}
```

File: CefSharp/ModelBinding/IInterceptor.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright © 2010-2017 The CefSharp Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+
+namespace CefSharp.ModelBinding
+{
+    /// <summary>
+    /// Intercepts methods execution
```
Comment (amaitland) [inline CefSharp/ModelBinding/IInterceptor.cs]: Same as above, needs a more descriptive comment.

File: CefSharp/ModelBinding/ModelBindingExtensions.cs [hunk 5]
```diff
@@ -36,7 +36,7 @@ public static T CreateInstance<T>(this Type type, bool nonPublic = false)
         /// <param name="nonPublic"><see langword="true"/> if a non-public constructor can be used, otherwise <see langword="false"/>.</param>
         public static object CreateInstance(this Type type, bool nonPublic = false)
         {
-            return CreateInstanceInternal(type, nonPublic);
+            return Activator.CreateInstance(type, nonPublic);
```
Comment (amaitland) [inline CefSharp/ModelBinding/ModelBindingExtensions.cs]: If your going to remove a large block of code, please add a comment to your commit description as to the reason please.

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs [hunk 6]
```diff
@@ -27,7 +39,12 @@ public BrowserTabView()
 
             browser.RequestHandler = new RequestHandler();
             browser.RegisterJsObject("bound", new BoundObject(), BindingOptions.DefaultBinder);
-            browser.RegisterAsyncJsObject("boundAsync", new AsyncBoundObject());
+            var bindingOptions = new BindingOptions()
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs]: Add a comment here as to what feature your demonstrating and it's purpose please.

**Comment by amaitland** [comment]
Also please make the `PR` title more descriptive, `Method interception` doesn't really explain much.

**Comment by joaompneves** [comment]
I understand that this might be a little bit over CefSharp responsibilities but implementing a generic decorator pattern in C# requires "heavy" machinery (eg: PostSharp) or dynamic AOP solutions that are usually clumsy and slow. Since there's already an "interception" mechanism for converting types (IBinder) I thought this could be a nice addition with very little overhead and complexity.

Are you willing to accept this PR after the requested changes are made?

**Comment by amaitland** [comment]
> Are you willing to accept this PR after the requested changes are made?

I'll think about it and get back to you.

**Comment by amaitland** [comment]
@joaompneves If your happy to make the following changes this can be merged.

- [ ] Rename `IInterceptor` to `IMethodInterceptor`
- [ ] Rename `BindingOptions.Interceptor` to `BindingOptions.MethodInterceptor`.
- [ ] Add detailed xml documentation to `IMethodInterceptor` interface and `BindingOptions.MethodInterceptor` property
- [ ] Previously code review comments

**Comment by joaompneves** [comment]
Made all the requested changes.

**Comment by amaitland** [comment]
Merged with follow up changes see https://github.com/cefsharp/CefSharp/commit/f5e2a938ec2d56553523baf8a25db539be1c487e

**Pull request merged**